libbpf-tools/hardirqs.c

```diff
@@ -0,0 +1,293 @@
+	struct irq_key lookup_key = {}, next_key;
+	struct hist hist;
+	int err;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		printf("hardirq = %s\n", next_key.name);
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	memset(&lookup_key, 0, sizeof(lookup_key));
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist: %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct hardirqs_bpf *obj;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = hardirqs_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	if (!env.count) {
+		obj->rodata->targ_dist = env.distributed;
+		obj->rodata->targ_ns = env.nanoseconds;
+	}
+
+	err = hardirqs_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	if (env.count) {
+		obj->links.handle__irq_handler =
+			bpf_program__attach(obj->progs.handle__irq_handler);
+		err = libbpf_get_error(obj->links.handle__irq_handler);
+		if (err) {
+			fprintf(stderr,
+				"failed to attach irq/irq_handler_entry: %s\n",
+				strerror(err));
+		}
+	} else {
+		obj->links.irq_handler_entry =
+			bpf_program__attach(obj->progs.irq_handler_entry);
+		err = libbpf_get_error(obj->links.irq_handler_entry);
+		if (err) {
+			fprintf(stderr,
+				"failed to attach irq_handler_entry: %s\n",
+				strerror(err));
+		}
+		obj->links.irq_handler_exit_exit =
+			bpf_program__attach(obj->progs.irq_handler_exit_exit);
+		err = libbpf_get_error(obj->links.irq_handler_exit_exit);
+		if (err) {
+			fprintf(stderr,
+				"failed to attach irq_handler_exit: %s\n",
+				strerror(err));
+		}
+	}
+
+	signal(SIGINT, sig_handler);
+
+	if (env.count)
+		printf("Tracing hard irq events... Hit Ctrl-C to end.\n");
+	else
+		printf("Tracing hard irq event time... Hit Ctrl-C to end.\n");
+
+	/* main: poll */
+	while (1) {
+		sleep(env.interval);
+		printf("\n");
+
+		if (env.timestamp) {
+			time(&t);
+			tm = localtime(&t);
+			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+			printf("%-8s\n", ts);
+		}
+
+		if (env.distributed)
+			err = print_hist(bpf_map__fd(obj->maps.hists));
+		else
+			err = print_count_or_latency(
+				bpf_map__fd(obj->maps.irqs));
```

Comment:
nit: pass `struct bpf_map *` as arguments to those functions. It will be shorter here and less error-prone, as you can't pass any random integer anymore.

Reply:
Thanks, will fix it after make sure the aboves

Reply:
Fixed, thanks!